controller/src/main/java/io/pravega/controller/server/rentention/AutoRetentionBucketService.java

```diff
@@ -0,0 +1,133 @@
+    private final ConcurrentMap<Stream, CompletableFuture> retentionFutureMap;
+    private final LinkedBlockingQueue<BucketChangeListener.StreamNotification> notifications;
+    private final AtomicBoolean stop;
+
+    public AutoRetentionBucketService(int bucketId, StreamMetadataStore streamMetadataStore,
+                                      StreamMetadataTasks streamMetadataTasks, ScheduledExecutorService executor) {
+        this.bucketId = bucketId;
+        this.streamMetadataStore = streamMetadataStore;
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.executor = executor;
+
+        this.notifications = new LinkedBlockingQueue<>();
+        this.stop = new AtomicBoolean(false);
+        this.retentionFutureMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    protected void doStart() {
+        Futures.await(RetryHelper.withIndefiniteRetriesAsync(() -> streamMetadataStore.getStreamsForBucket(bucketId, executor)
+                .thenAccept(streams -> retentionFutureMap.putAll(streams.stream()
+                        .map(s -> {
+                            String[] splits = s.split("/");
+                            return new StreamImpl(splits[0], splits[1]);
+                        })
+                        .collect(Collectors.toMap(s -> s, this::getStreamRetentionFuture))
+                )),
+                e -> log.warn("exception thrown getting streams for bucket {}, e = {}", bucketId, e), executor)
+                .thenAccept(x -> {
+                    log.info("streams collected for the bucket, registering for change notification and starting loop for processing notifications");
+                    streamMetadataStore.registerBucketChangeListener(bucketId, this);
+                    Futures.loop(stop::get, this::processNotification, executor);
+                }));
+        notifyStarted();
+    }
+
+    private CompletableFuture<Void> processNotification() {
+        return CompletableFuture.runAsync(() -> {
+            StreamNotification notification = Exceptions.handleInterrupted(notifications::take);
+            StreamImpl stream = new StreamImpl(notification.getScope(), notification.getStream());
+
+            switch (notification.getType()) {
+                case StreamAdded:
+                    retentionFutureMap.putIfAbsent(stream, getStreamRetentionFuture(stream));
+                    break;
+                case StreamRemoved:
+                    retentionFutureMap.remove(stream).cancel(true);
+                    break;
+                case StreamUpdated:
+                    // if retention has been disabled then cancel the future.
+                    // if retention has been enabled then add the future if it does not exist.
+                    // For now we will do nothing.
+                    break;
+                case ConnectivityError:
+                    // TODO: refresh retention future map as this means we can miss events
```

Comment:
I checked with @shrids that curator recipe for pathcache makes sure we dont miss events in case of connection failures. 
I need to simulate that in a unit test to verify if that is the case or not. 
If its true then this todo is not valid and can be safely removed.

Reply:
Ok, update this thread once you know.

Reply:
Description for what happens when we get ConnectionLost error. 
`        /**
         * Called when the connection has changed to {@link ConnectionState#LOST}
         *
         * This is exposed so that users of the class can be notified of issues that *might* affect normal operation.
         * The PathChildrenCache is written such that listeners are not expected to do anything special on this
         * event, except for those people who want to cause some application-specific logic to fire when this occurs.
         * While the connection is down, the PathChildrenCache will continue to have its state from before it lost
         * the connection and after the connection is restored, the PathChildrenCache will emit normal child events
         * for all of the adds, deletes and updates that happened during the time that it was disconnected.
         */
`

So we dont need to do anything here